pkg/controller/node/nodecontroller.go

```diff
@@ -561,7 +561,9 @@ func (nc *NodeController) Run() {
 						} else {
 							glog.V(4).Info("Made sure that Node %v has no %v Taint", value.Value, oppositeTaint)
 						}
-						return true, 0
+
+						// remove the node from the queue after the adding/removal of taint is successful
+						return nc.zoneNotReadyOrUnreachableTainer[utilnode.GetZoneKey(node)].Remove(node.Name), 0
```

Comment:
Current node entry is in the list with condition NotReady. When the node condition changes from notready to unknown, node with updated condition of Unknown is not added to the list because old node entry (with condition NotReady) is still there. Due to this, in the code above, condition=Unknown is never reached and consequently node taint is not updated to Unreachable and NotReady taint stays.

To fix this logic, old node entry is deleted, so that new node entry with updated condition could be added and right action can be taken.